docs/proposals/receiving_metadata.md

```diff
@@ -0,0 +1,49 @@
+
+## Potential Solutions
+Before we delve into the solutions, let's set a baseline about how the data is received. This applies almost equally for the two.
+
+Metadata from Prometheus is sent in the same `WriteRequest` proto message that the samples use. It is part of a different field (#3 given #2 is already used), the data is a set identified by the metric name - that means it is aggregated across targets, and finally it is sent all at once. It is also important to note that this current process is an intermediary step. Eventually, metadata in a request will be sent alongside samples and only for those included. The solutions proposed, take this nuance into account to avoid coupling between the current and future state of Prometheus, and hopefully do something now that also works for the future. 
+
+As a reference, these are some key numbers regarding the size (and send timings) of the data at hand from our clusters at GL:
+
+- On average, metadata (a combination of `HELP`, `TYPE`, `UNIT` and `METRIC_NAME`) is ~55 bytes uncompressed.
+- at GL, on an instance with about 2.6M active series, we hold ~1241 unique metrics in total.
+- with that, we can assume that on a worst-case scenario the metadata set for that instance is ~68 kilobytes uncompressed.
+- by default, this data is only propagated once every minute (aligning with the default scrape interval), but this can be adjusted.
+- Finally, what this gives us is a baseline worst-case scenario formula for the data to store per tenant: `~68KB * Replication Factor * # of Instances`. Keeping in mind that typically, there's a very high overlap of metadata across instances, and we plan to deduplicate in the ingesters.
+
+### Write Path
+
+1. Store the metadata directly from the distributors into a cache (e.g. Memcached)
+
+Since metadata is received all at once, we could directly store into an external cache using the tenant ID as a key, and still, avoid a read-modify-write.  However, a very common use case of Cortex is to have multiple Prometheus sending data for the same tenant ID. This complicates things, as it adds a need to have an intermediary merging phase and thus making a read-modify-write inevitable.
+
+2. Keep metadata in memory within the ingesters
+
+Similarly to what we do with sample data, we can keep the metadata in-memory in the ingesters and apply similar semantics. I propose to use the tenant ID as a hash key, distribute it to the ingesters (taking into account the replication factor), using a hash map to keep a set of the metadata across all instances for a single tenant, and implement a configurable time-based purge process to deal with metadata churn. Given, we need to ensure fair-use we also propose implementing limits for both the number of metadata entries we can receive and the size of a single entry.
```

Comment:
Can we do `tenant ID` + `metric_name` for the hash key? There are clusters where there is only one tenant, and they can hit 30-50Mil metrics. While its not really an issue (memory is small at that scale too!), I wonder what the downside of spreading it across ingesters is.

Reply:
> There are clusters where there is only one tenant, and they can hit 30-50Mil metrics
Are we talking about 30-50k metrics or series?

We basically have two access patterns that we need to support as part of the API, give me all metrics with their metadata and give me all metadata of a single metrics. At the moment, the most common one is the former. If my understanding is correct, if we use `tenant ID + metric name` we'd have to look across all ingesters to support the most used pattern.

Reply:
Yes, but we query all ingesters for _all queries_ anyways. Might be a premature optimisation though. We can always split it across ingesters in the future (if its actually a bottleneck), but lets make a note of it.

Reply:
I thought we [only queried all of the ingesters](https://github.com/cortexproject/cortex/blob/master/docs/architecture.md#hashing) if we hashed by labels which is the non-default option. If the documentation is outdated, then I see no reason to not use the metric name.

Reply:
Haha, yes thats true. But for large installations, we use shard-by-all-labels.